File: src/main/providers/translate/baiduAi.ts
import { defineTranslateProvider } from '@main/providers/translate';
import crypto from 'crypto';
import querystring from 'querystring';
import fetch from 'electron-fetch';

export default defineTranslateProvider({
    id: '百度AI开放平台',
    description: '百度AI开放平台机器翻译API，详见 https://api.fanyi.baidu.com \n此翻译器会向 api.fanyi.baidu.com 发送数据',
    optionsSchema: {
        enable: Boolean,
        apiConfig: {
            appid: [String, null] as const,
            key: [String, null] as const,
            fromLanguage: String,
            toLanguage: String
        }
    },
    defaultOptions: {
        enable: true,
        apiConfig: {
            appid: null,
            key: null,
            fromLanguage: 'jp',
            toLanguage: 'zh'
        }
    },
    optionsDescription: {
        enable: '启用',
        apiConfig: {
            appid: { readableName: 'APP ID', description: '可在 https://api.fanyi.baidu.com/api/trans/product/desktop?req=developer 获取' },
            key: { readableName: 'Key', description: '可在 https://api.fanyi.baidu.com/api/trans/product/desktop?req=developer 获取' },
            fromLanguage: '源语言',
            toLanguage: '目标语言'
        }
    },
    data() {
        return null;
    }
}, {
    init() { },
    isReady() {
        return this.enable &&
            this.apiConfig.appid !== null &&
            this.apiConfig.key !== null;
    },
    async translate(text) {
        const salt = (new Date()).getTime();
        const str1 = this.apiConfig.appid + text + salt + this.apiConfig.key;
        const md5 = crypto.createHash('md5');
        const sign = md5.update(str1).digest('hex');
        return fetch('https://fanyi-api.baidu.com/api/trans/vip/translate?' + querystring.stringify({
            q: text,
            appid: this.apiConfig.appid,
            salt,
            from: this.apiConfig.fromLanguage,
            to: this.apiConfig.toLanguage,
            sign
        }))
            .then(res => res.json())
            .then(res => {
                if (!res.trans_result) throw res;
                return res.trans_result.map((i: any) => i.dst).join('\n');
            });
    }
});

File: src-tauri/src/providers/translate/baidu_ai.rs
//! Baidu Translate API provider (百度AI开放平台).
//!
//! GET https://fanyi-api.baidu.com/api/trans/vip/translate
//! Signing: MD5(appid + text + salt + key) as lowercase hex.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::TranslateProvider;
use crate::crypto::md5_hex;
use crate::schema::{AmeOptions, JsonSchema};

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema, AmeOptions)]
#[serde(rename_all = "camelCase")]
pub struct BaiduOptions {
    #[ame(desc = "启用")]
    #[serde(default)]
    pub enable: bool,
    #[ame]
    #[serde(default)]
    pub api_config: BaiduApiConfig,
}

#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema, AmeOptions)]
#[serde(rename_all = "camelCase")]
pub struct BaiduApiConfig {
    #[ame(
        readable = "APP ID",
        desc = "可在 https://fanyi-api.baidu.com/api/trans/product/desktop?req=developer 获取"
    )]
    #[serde(default)]
    pub appid: Option<String>,
    #[ame(
        readable = "Key",
        desc = "可在 https://fanyi-api.baidu.com/api/trans/product/desktop?req=developer 获取"
    )]
    #[serde(default)]
    pub key: Option<String>,
    #[ame(desc = "源语言")]
    #[serde(default = "default_from")]
    pub from_language: String,
    #[ame(desc = "目标语言")]
    #[serde(default = "default_to")]
    pub to_language: String,
}

fn default_from() -> String {
    "jp".into()
}
fn default_to() -> String {
    "zh".into()
}

impl Default for BaiduApiConfig {
    fn default() -> Self {
        Self {
            appid: None,
            key: None,
            from_language: default_from(),
            to_language: default_to(),
        }
    }
}

impl Default for BaiduOptions {
    fn default() -> Self {
        Self {
            enable: true,
            api_config: BaiduApiConfig {
                appid: None,
                key: None,
                from_language: default_from(),
                to_language: default_to(),
            },
        }
    }
}

pub struct Baidu {
    pub options: BaiduOptions,
    client: reqwest::Client,
}

impl Baidu {
    pub fn new(options: BaiduOptions) -> Self {
        Self {
            options,
            client: reqwest::Client::new(),
        }
    }
}

impl TranslateProvider for Baidu {
    fn id(&self) -> &str {
        "百度AI开放平台"
    }

    fn options_schema() -> Value {
        <BaiduOptions as AmeOptions>::schema()
    }

    fn default_options() -> Value {
        serde_json::to_value(BaiduOptions::default()).unwrap()
    }

    fn options_description() -> Value {
        <BaiduOptions as AmeOptions>::description()
    }

    fn enabled(&self) -> bool {
        self.options.enable
            && self.options.api_config.appid.is_some()
            && self.options.api_config.key.is_some()
    }

    async fn translate(&self, text: String) -> Result<String, String> {
        let cfg = &self.options.api_config;
        let appid = cfg.appid.clone().unwrap_or_default();
        let key = cfg.key.clone().unwrap_or_default();
        let salt = chrono::Utc::now().timestamp_millis().to_string();
        let sign = md5_hex(&format!("{appid}{text}{salt}{key}"));

        let resp = self
            .client
            .get("https://fanyi-api.baidu.com/api/trans/vip/translate")
            .header("connection", "keep-alive")
            .query(&[
                ("q", text.as_str()),
                ("appid", appid.as_str()),
                ("salt", salt.as_str()),
                ("from", cfg.from_language.as_str()),
                ("to", cfg.to_language.as_str()),
                ("sign", sign.as_str()),
            ])
            .send()
            .await
            .map_err(|e| e.to_string())?;
        let json: Value = resp.json().await.map_err(|e| e.to_string())?;

        if let Some(arr) = json["trans_result"].as_array() {
            let result: Vec<&str> = arr.iter().filter_map(|i| i["dst"].as_str()).collect();
            if !result.is_empty() {
                return Ok(result.join("\n"));
            }
        }
        Err(json["error_msg"]
            .as_str()
            .unwrap_or("Unknown Baidu error")
            .to_string())
    }
}
